pkg/controller/snapshot/network/controller.go

```diff
@@ -137,8 +137,8 @@ func (r *Reconciler) reconcileRunningMark(snapshot *networksnaptypes.NetworkSnap
 // createDeviceSnapshots marks NetworkChanges for deletion and creates device snapshots
 func (r *Reconciler) createDeviceSnapshots(snapshot *networksnaptypes.NetworkSnapshot) (bool, error) {
 	// Iterate through network changes
-	deviceChanges := make(map[device.ID]networkchangetypes.Index)
-	deviceMaxChanges := make(map[device.ID]networkchangetypes.Index)
+	deviceChanges := make(map[device.ID]map[string]networkchangetypes.Index)
```

Comment:
So, this could be implemented with a device+version key or a sub map as it is now. I think I’ll change to the former.

But as the controller iterates through the global `NetworkChange` log, these two maps keep track of:
* `deviceChanges` - The last seen index for each device+version that meets the criteria for the snapshot
* `deviceMaxChanges` The highest index for each device+version that meets the criteria for the snapshot

So, as the controller iterates through changes, when it sees a change that meets the snapshot criteria, it records the index of each device+version. The first time it encounters a change that _doesn’t_ meet the criteria for the snapshot, it records the last seen index for each device in the change as the max snapshot index.

So, `deviceChanges` is a map that’s just used to build the `deviceMaxChanges` map, and the `deviceMaxChanges` map contains the index at which to take the snapshot for each device+version. The max index for each device is used to ensure we retain changes we want to retain (e.g. pending changes) and is passed to the `DeviceSnapshot` controllers.

But you’re right. This should probably be documented in the code.

Reply:
+1 to documenting it - there could be easily someone as dumb as me comes along again in the future and can't figure it out - or it could be my future self :wink: